functions/src/index.ts

```diff
@@ -0,0 +1,128 @@
+
+import * as admin from "firebase-admin";
+import * as functions from "firebase-functions";
+import { PortalFirebaseJWTClaims } from "./portal-types";
+
+// set to true to enable additional logging
+const DEBUG=true;
+
+// Start writing Firebase Functions
+// https://firebase.google.com/docs/functions/typescript
+
+const app = admin.initializeApp({
+  // credential: admin.credential.applicationDefault(),
+  databaseURL: 'https://collaborative-learning-ec215.firebaseio.com/'
+});
+const firebase = app.database();
+const firestore = app.firestore();
+
+export const helloWorld = functions.https.onRequest((request, response) => {
+  functions.logger.info("Hello logs (onRequest)!", {structuredData: true});
+  response.send("Hello from Firebase onRequest!");
+});
+
+interface IHelloCallerParams {
+  caller?: string;
+}
+export const helloCaller = functions.https.onCall((data: IHelloCallerParams, context) => {
+  const { caller } = data;
+  const message = `Hello ${caller} from Firebase onCall!`;
+  functions.logger.info(`LOG: ${message}`, {structuredData: true});
+  return message;
+});
+
+// contents of context.auth?.token in firebase callable functions
+type IDecodedIdToken = admin.auth.DecodedIdToken & Partial<PortalFirebaseJWTClaims>;
+
+// relevant contents of multi-class support document in firestore
+interface IFirestoreSupportPublication {
+  classes: string[];
+  context_id: string;   // classHash
+  platform_id: string;  // portal, e.g. "learn.concord.org"
+}
+
+// parameters passed to getImageData() function by client
+interface IGetImageDataParams {
+  url: string;
+  appMode: string;
+  demoName: string;
+  portal: string;
+  classHash: string;
+  classPath: string;
+  type?: string;
+  key?: string;
+}
+export const getImageData = functions.https.onCall(async (data: IGetImageDataParams, context) => {
+  // these values are provided by the client and so should be considered potentially suspect
+  const { url, appMode, demoName, portal: _portal, classHash: _classHash, classPath, type, key } = data;
+  // these claim values come from the JWT and so should be considered verified/authenticated
+  // they won't be present in dev/demo/qa modes, however, where security isn't a concern
+  const claims: IDecodedIdToken | undefined = context.auth?.token;
+  const portal: string = claims?.platform_id ||
+                          (_portal === "demo" ? demoName : _portal) ||
+                          "";
```

Comment:
So my understanding is that an incoming request can make the portal be anything it wants if there aren't claims. But in that case hasValidClaims will be false. So then appMode has to be something other than 'authed'. 

Is it necessary to allow any random appMode, or could that be restricted so anonymous clients can't create new top level keys?

Reply:
I added the appMode restrictions while adding the unit tests.